File: day5/typescript/part_1.ts
import * as fs from 'fs';

type Rule = [number, number];
type Update = number[];

function parseInput(filePath: string): [Rule[], Update[]] {
    const lines = fs.readFileSync(filePath, 'utf-8').trim().split('\n');
    const rules: Rule[] = [];
    const updates: Update[] = [];
    let section = 0;

    for (const line of lines) {
        if (!line.trim()) {
            section++;
            continue;
        }
        if (section === 0) {
            rules.push(line.split('|').map(Number) as Rule);
        } else {
            updates.push(line.split(',').map(Number));
        }
    }

    return [rules, updates];
}

function isUpdateOrdered(update: Update, rules: Rule[]): boolean {
    const pageIndices = new Map(update.map((page, i) => [page, i]));

    for (const [a, b] of rules) {
        if (pageIndices.has(a) && pageIndices.has(b)) {
            if (pageIndices.get(a)! > pageIndices.get(b)!) {
                return false;
            }
        }
    }
    return true;
}

const [rules, updates] = parseInput('input.txt');
const middleSum = updates
    .filter(update => isUpdateOrdered(update, rules))
    .reduce((sum, update) => sum + update[Math.floor(update.length / 2)], 0);

console.log(middleSum);

File: day5/rust/part_1.rs
use std::collections::HashMap;
use std::fs::read_to_string;

type Rule = (i32, i32);
type Update = Vec<i32>;

fn parse_input(file_path: &str) -> (Vec<Rule>, Vec<Update>) {
    let content = read_to_string(file_path).unwrap();
    let mut rules = Vec::new();
    let mut updates = Vec::new();
    let mut section = 0;

    for line in content.lines() {
        let line = line.trim();
        if line.is_empty() {
            section += 1;
            continue;
        }

        if section == 0 {
            let parts: Vec<i32> = line.split('|')
                .map(|s| s.parse().unwrap())
                .collect();
            rules.push((parts[0], parts[1]));
        } else {
            let update: Vec<i32> = line.split(',')
                .map(|s| s.parse().unwrap())
                .collect();
            updates.push(update);
        }
    }

    (rules, updates)
}

fn is_update_ordered(update: &[i32], rules: &[Rule]) -> bool {
    let page_indices: HashMap<i32, usize> = update.iter()
        .enumerate()
        .map(|(i, &page)| (page, i))
        .collect();

    for &(a, b) in rules {
        if let (Some(&i1), Some(&i2)) = (page_indices.get(&a), page_indices.get(&b)) {
            if i1 > i2 {
                return false;
            }
        }
    }
    true
}

fn main() {
    let (rules, updates) = parse_input("input.txt");
    let middle_sum: i32 = updates.iter()
        .filter(|update| is_update_ordered(update, &rules))
        .map(|update| update[update.len() / 2])
        .sum();

    println!("{}", middle_sum);
}
